File: scripts/audit-response-logs.py
#!/usr/bin/env python3
import argparse
import json
import re
from collections import Counter, defaultdict
from pathlib import Path

UI_BREAK_PATTERNS = {
    "boxed": re.compile(r"\\boxed\s*\{"),
    "bracket_display": re.compile(r"\\\[|\\\]"),
    "paren_inline": re.compile(r"\\\(|\\\)"),
    "single_dollar_line": re.compile(r"(?m)^\s*\$(?!\$)\s*$"),
    "raw_latex_outside_display": re.compile(
        r"\\(?:frac|sqrt|int|sum|lim|begin|left|right|cdot)\b"
    ),
}

NEMANJA_MARKERS = [
    "kalk",
    "so",
    "now",
    "remember",
    "board",
    "what do we",
    "there we go",
    "that's it",
    "keep in mind",
    "does that make sense",
]

FAILURE_TAXONOMY = {
    "SAN": "Sanitizer Leak",
    "DISC": "Discrepancy",
    "RAG": "RAG Route",
    "UI": "UI Breaking",
    "LOGIC": "Math/Answer Logic",
    "REQ": "Request Failure",
    "UNK": "Uncategorized",
}


def code_for_failure(failure: str) -> str:
    text = failure.lower()
    if any(token in text for token in ["prose inside display math", "raw latex outside", "broken step", "placeholder"]):
        return "SAN"
    if any(token in text for token in ["boxed", "bracket", "paren", "single_dollar", "unsupported", "unbalanced", "empty_display"]):
        return "UI"
    if any(token in text for token in ["mismatch", "final answer", "logic"]):
        return "DISC"
    if "rag" in text or "grounding" in text or "lecture" in text:
        return "RAG"
    return "UNK"


def strip_display_math(text: str) -> str:
    return re.sub(r"\$\$[\s\S]*?\$\$", "", text)


def final_answer(text: str) -> str:
    match = re.search(r"(?:##\s*)?Final Answer\s*:?\s*([\s\S]*)", text, re.I)
    if not match:
        return ""
    answer = match.group(1)
    math_blocks = re.findall(r"\$\$([\s\S]*?)\$\$", answer)
    if math_blocks:
        answer = math_blocks[-1]
    else:
        first_paragraph = answer.split("\n\n", 1)[0]
        trailing_value = re.search(
            r"(?:is|equals|gives|result\s+is)\s*:?\s*(-?\d+(?:\.\d+)?)\s*\.?\s*$",
            first_paragraph,
            re.I,
        )
        equation = re.search(
            r"(?:[a-z]\s*=\s*)?(?:[a-z]\s*)?[a-z]\s*[+\-]\s*\d+(?:\.\d+)?",
            first_paragraph,
            re.I,
        )
        numbers = re.findall(r"-?\d+(?:\.\d+)?", first_paragraph)
        if trailing_value:
            answer = trailing_value.group(1)
        elif equation:
            answer = equation.group(0)
        elif numbers and len(first_paragraph) > 30:
            answer = numbers[-1]
        else:
            answer = first_paragraph
    answer = re.sub(r"\$\$", "", answer)
    answer = re.sub(r"\s+", " ", answer).strip()
    return answer[:500]


def normalize_answer(answer: str) -> str:
    normalized = answer.lower()
    normalized = normalized.replace("\\left", "").replace("\\right", "")
    normalized = normalized.replace("\\,", "").replace(" ", "")
    normalized = normalized.replace("{", "").replace("}", "")
    normalized = normalized.replace("\\quad", "")

    matrix_match = re.search(r"(?:[a-z]{1,3}=)?(\\beginbmatrix[\s\S]*?\\endbmatrix)", normalized)
    if matrix_match:
        return matrix_match.group(1)

    y_equation = re.search(r"y=x[+-]\d+(?:\.\d+)?", normalized)
    if y_equation:
        return y_equation.group(0)

    normalized = normalized.replace("f'(x)=", "")
    normalized = normalized.replace("y=", "")

    ordered_triple = re.search(r"(?:x=)?(-?\d+(?:\.\d+)?),?(?:y=)?(-?\d+(?:\.\d+)?),?(?:z=)?(-?\d+(?:\.\d+)?)", normalized)
    if ordered_triple and ("x=" in normalized or "(x,y,z)" in normalized):
        return ",".join(ordered_triple.groups())

    if "mean" in normalized and ("variance" in normalized or "standarddeviation" in normalized):
        return ",".join(re.findall(r"-?\d+(?:\.\d+)?", normalized))

    if "z-score" in normalized or normalized.startswith("z="):
        decimals = re.findall(r"-?\d+\.\d+", normalized)
        if decimals:
            return decimals[-1]
        rhs = re.search(r"z=(-?\d+(?:\.\d+)?)", normalized)
        if rhs:
            return rhs.group(1)

    if "sum" in normalized or "\\sum" in normalized:
        numbers = re.findall(r"-?\d+(?:\.\d+)?", normalized)
        if numbers:
            return numbers[-1]

    value_phrase = re.search(r"(?:value|price|finalprice|equals|is)(?:[^-0-9]*)(-?\d+(?:\.\d+)?)\.?$", normalized)
    if value_phrase:
        return value_phrase.group(1)

    simple_rhs = re.search(r"=(-?\d+(?:\.\d+)?)\.?$", normalized)
    if simple_rhs and normalized.count("=") == 1:
        return simple_rhs.group(1)

    return normalized


def ui_breaks(text: str):
    outside = strip_display_math(text)
    failures = []
    for name, pattern in UI_BREAK_PATTERNS.items():
        target = outside if name == "raw_latex_outside_display" else text
        if pattern.search(target):
            failures.append(name)
    if text.count("$$") % 2:
        failures.append("unbalanced_display_fences")
    if any(not match.strip() for match in re.findall(r"\$\$([\s\S]*?)\$\$", text)):
        failures.append("empty_display_block")
    return failures


def add_failure(code_counter, detail_counter, examples, code, detail, entry):
    code_counter[code] += 1
    detail_counter[code][detail] += 1
    if len(examples[code]) < 8:
        examples[code].append(
            {
                "id": entry.get("id"),
                "mode": entry.get("mode"),
                "category": entry.get("category"),
                "prompt": entry.get("prompt"),
                "detail": detail,
            }
        )


def persona_density(text: str) -> int:
    lower = text.lower()
    return sum(lower.count(marker) for marker in NEMANJA_MARKERS)


def grounding_hits(entry) -> int:
    expected = entry.get("expectedGroundingKeywords") or []
    if not expected:
        return 0
    lower = (entry.get("output") or "").lower()
    return sum(1 for keyword in expected if keyword.lower() in lower)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("path", nargs="?", default="scripts/response_logs.json")
    parser.add_argument("--json", action="store_true", help="print machine-readable JSON")
    args = parser.parse_args()

    path = Path(args.path)
    entries = json.loads(path.read_text(encoding="utf-8"))

    ui_counter = Counter()
    failure_counter = Counter()
    failure_details = defaultdict(Counter)
    failure_examples = defaultdict(list)
    persona_scores = defaultdict(list)
    grounding_missing = []
    grouped = defaultdict(dict)

    for entry in entries:
        output = entry.get("output") or ""
        for failure in ui_breaks(output):
            ui_counter[failure] += 1
            add_failure(failure_counter, failure_details, failure_examples, code_for_failure(failure), failure, entry)
        for failure in entry.get("failures") or []:
            add_failure(failure_counter, failure_details, failure_examples, code_for_failure(failure), failure, entry)
        mode = entry.get("mode", "unknown")
        persona_scores[mode].append(persona_density(output))
        if entry.get("expectedGroundingKeywords") and grounding_hits(entry) == 0:
            grounding_missing.append(entry.get("id") or entry.get("prompt"))
            add_failure(
                failure_counter,
                failure_details,
                failure_examples,
                "RAG",
                "missing expected grounding keyword",
                entry,
            )
        key = (entry.get("pass"), entry.get("prompt"))
        grouped[key][mode] = normalize_answer(final_answer(output))

    mismatches = []
    by_category = Counter()
    category_totals = Counter()
    for key, modes in grouped.items():
        pure = modes.get("pure")
        if not pure:
            continue
        for mode, answer in modes.items():
            if mode == "pure" or not answer:
                continue
            category = next(
                (
                    entry.get("category", "uncategorized")
                    for entry in entries
                    if (entry.get("pass"), entry.get("prompt")) == key
                ),
                "uncategorized",
            )
            category_totals[category] += 1
            if answer != pure:
                mismatches.append({"key": key, "mode": mode, "pure": pure, "other": answer, "category": category})
                by_category[category] += 1
                add_failure(
                    failure_counter,
                    failure_details,
                    failure_examples,
                    "DISC",
                    "Nemanja/Pure Logic final answer mismatch",
                    {
                        "id": None,
                        "mode": mode,
                        "category": category,
                        "prompt": key[1],
                    },
                )

    mismatch_rate = (len(mismatches) / max(1, sum(category_totals.values()))) * 100
    persona_summary = {
        mode: {
            "count": len(scores),
            "average": round(sum(scores) / max(1, len(scores)), 3),
            "min": min(scores) if scores else 0,
            "max": max(scores) if scores else 0,
        }
        for mode, scores in persona_scores.items()
    }

    summary = {
        "entries": len(entries),
        "ui_breaks": dict(ui_counter),
        "failure_taxonomy": FAILURE_TAXONOMY,
        "failures_by_code": {
            code: {
                "label": FAILURE_TAXONOMY.get(code, "Uncategorized"),
                "count": failure_counter.get(code, 0),
                "details": dict(failure_details.get(code, {})),
                "examples": failure_examples.get(code, []),
            }
            for code in FAILURE_TAXONOMY
            if failure_counter.get(code, 0)
        },
        "mismatch_count": len(mismatches),
        "mismatch_rate_percent": round(mismatch_rate, 3),
        "mismatch_by_category": dict(by_category),
        "persona_density": persona_summary,
        "grounding_missing_count": len(grounding_missing),
        "grounding_missing_examples": grounding_missing[:20],
        "requires_targeted_expansion": [
            category
            for category, total in category_totals.items()
            if total and (by_category[category] / total) > 0.03
        ],
    }

    if args.json:
        print(json.dumps(summary, indent=2))
    else:
        print(json.dumps(summary, indent=2))
        if summary["requires_targeted_expansion"]:
            print("\nTargeted expansion needed for:", ", ".join(summary["requires_targeted_expansion"]))


if __name__ == "__main__":
    main()

File: scripts/audit-response-logs.mjs
import fs from "node:fs";

const FAILURE_TAXONOMY = {
  SAN: "Sanitizer Leak",
  DISC: "Discrepancy",
  RAG: "RAG Route",
  UI: "UI Breaking",
  LOGIC: "Math/Answer Logic",
  REQ: "Request Failure",
  UNK: "Uncategorized",
};

const UI_BREAK_PATTERNS = {
  boxed: /\\boxed\s*\{/,
  bracket_display: /\\\[|\\\]/,
  paren_inline: /\\\(|\\\)/,
  single_dollar_line: /^\s*\$(?!\$)\s*$/m,
  raw_latex_outside_display:
    /\\(?:frac|sqrt|int|sum|lim|prod|begin|end|left|right|cdot|times|text|operatorname|ln|log|sin|cos|tan|sec|csc|cot|arcsin|arccos|arctan|pi|theta|alpha|beta|gamma|delta|lambda|mu|sigma|infty|to|leq|geq|neq|le|ge|pm|nabla|partial|vec|bar|hat|overline|underline)\b/,
  invalid_backslash_number: /(?<!\\)\\[0-9]/,
};

const NEMANJA_MARKERS = [
  "kalk",
  "so",
  "now",
  "remember",
  "board",
  "what do we",
  "there we go",
  "that's it",
  "keep in mind",
  "does that make sense",
];

function codeForFailure(failure) {
  const text = String(failure).toLowerCase();
  if (
    [
      "prose inside display math",
      "raw latex outside",
      "raw_latex_outside",
      "raw latex command",
      "broken step",
      "placeholder",
    ].some((token) => text.includes(token))
  ) {
    return "SAN";
  }
  if (
    [
      "boxed",
      "bracket",
      "paren",
      "single_dollar",
      "unsupported",
      "unbalanced",
      "empty_display",
      "invalid_backslash",
    ].some((token) => text.includes(token))
  ) {
    return "UI";
  }
  if (["mismatch", "final answer", "logic"].some((token) => text.includes(token))) {
    return "DISC";
  }
  if (["rag", "grounding", "lecture"].some((token) => text.includes(token))) {
    return "RAG";
  }
  return "UNK";
}

function stripDisplayMath(text) {
  return String(text).replace(/\$\$[\s\S]*?\$\$/g, "");
}

function finalAnswer(text) {
  const match = String(text).match(/(?:##\s*)?Final Answer\s*:?\s*([\s\S]*)/i);
  if (!match) return "";

  let answer = match[1];
  const mathBlocks = [...answer.matchAll(/\$\$([\s\S]*?)\$\$/g)].map((item) => item[1]);
  if (mathBlocks.length) {
    answer = mathBlocks.at(-1) ?? "";
  } else {
    const firstParagraph = answer.split(/\n\s*\n/, 1)[0] ?? "";
    const trailingValue = firstParagraph.match(
      /(?:is|equals|gives|result\s+is)\s*:?\s*(-?\d+(?:\.\d+)?)\s*\.?\s*$/i,
    );
    const equation = firstParagraph.match(
      /(?:[a-z]\s*=\s*)?(?:[a-z]\s*)?[a-z]\s*[+\-]\s*\d+(?:\.\d+)?/i,
    );
    const numbers = [...firstParagraph.matchAll(/-?\d+(?:\.\d+)?/g)].map((item) => item[0]);

    if (trailingValue) answer = trailingValue[1];
    else if (equation) answer = equation[0];
    else if (numbers.length && firstParagraph.length > 30) answer = numbers.at(-1) ?? "";
    else answer = firstParagraph;
  }

  return answer.replace(/\$\$/g, "").replace(/\s+/g, " ").trim().slice(0, 500);
}

function normalizeAnswer(answer) {
  let normalized = String(answer).toLowerCase();
  normalized = normalized
    .replaceAll("\\left", "")
    .replaceAll("\\right", "")
    .replaceAll("\\,", "")
    .replace(/\s+/g, "")
    .replace(/[{}]/g, "")
    .replaceAll("\\quad", "");

  const matrixMatch = normalized.match(/(?:[a-z]{1,3}=)?(\\beginbmatrix[\s\S]*?\\endbmatrix)/);
  if (matrixMatch) return matrixMatch[1];

  const yEquation = normalized.match(/y=x[+-]\d+(?:\.\d+)?/);
  if (yEquation) return yEquation[0];

  normalized = normalized.replaceAll("f'(x)=", "").replaceAll("y=", "");

  const orderedTriple = normalized.match(
    /(?:x=)?(-?\d+(?:\.\d+)?),?(?:y=)?(-?\d+(?:\.\d+)?),?(?:z=)?(-?\d+(?:\.\d+)?)/,
  );
  if (orderedTriple && (normalized.includes("x=") || normalized.includes("(x,y,z)"))) {
    return orderedTriple.slice(1, 4).join(",");
  }

  if (normalized.includes("mean") && (normalized.includes("variance") || normalized.includes("standarddeviation"))) {
    return [...normalized.matchAll(/-?\d+(?:\.\d+)?/g)].map((item) => item[0]).join(",");
  }

  if (normalized.includes("z-score") || normalized.startsWith("z=")) {
    const decimals = [...normalized.matchAll(/-?\d+\.\d+/g)].map((item) => item[0]);
    if (decimals.length) return decimals.at(-1) ?? "";
    const rhs = normalized.match(/z=(-?\d+(?:\.\d+)?)/);
    if (rhs) return rhs[1];
  }

  if (normalized.includes("sum") || normalized.includes("\\sum")) {
    const numbers = [...normalized.matchAll(/-?\d+(?:\.\d+)?/g)].map((item) => item[0]);
    if (numbers.length) return numbers.at(-1) ?? "";
  }

  const valuePhrase = normalized.match(/(?:value|price|finalprice|equals|is)(?:[^-0-9]*)(-?\d+(?:\.\d+)?)\.?$/);
  if (valuePhrase) return valuePhrase[1];

  const latexFractionRhs = normalized.match(/=(\\frac-?\d+(?:\.\d+)?-?\d+(?:\.\d+)?)\.?$/);
  if (latexFractionRhs && [...normalized.matchAll(/=/g)].length === 1) return latexFractionRhs[1];

  const simpleRhs = normalized.match(/=(-?\d+(?:\.\d+)?)\.?$/);
  if (simpleRhs && [...normalized.matchAll(/=/g)].length === 1) return simpleRhs[1];

  return normalized;
}

function uiBreaks(text) {
  const output = String(text);
  const outside = stripDisplayMath(output);
  const failures = [];

  for (const [name, pattern] of Object.entries(UI_BREAK_PATTERNS)) {
    const target = name === "raw_latex_outside_display" ? outside : output;
    if (pattern.test(target)) failures.push(name);
  }

  if ((output.match(/\$\$/g) ?? []).length % 2) failures.push("unbalanced_display_fences");
  for (const match of output.matchAll(/\$\$([\s\S]*?)\$\$/g)) {
    if (!match[1].trim()) failures.push("empty_display_block");
  }

  return failures;
}

function addFailure(codeCounter, detailCounter, examples, code, detail, entry) {
  codeCounter[code] = (codeCounter[code] ?? 0) + 1;
  detailCounter[code] ??= {};
  detailCounter[code][detail] = (detailCounter[code][detail] ?? 0) + 1;
  examples[code] ??= [];
  if (examples[code].length < 8) {
    examples[code].push({
      id: entry.id,
      mode: entry.mode,
      category: entry.category,
      prompt: entry.prompt,
      detail,
    });
  }
}

function personaDensity(text) {
  const lower = String(text).toLowerCase();
  return NEMANJA_MARKERS.reduce((count, marker) => {
    return count + lower.split(marker).length - 1;
  }, 0);
}

function groundingHits(entry) {
  const expected = entry.expectedGroundingKeywords ?? [];
  if (!expected.length) return 0;
  const lower = String(entry.output ?? "").toLowerCase();
  return expected.filter((keyword) => lower.includes(String(keyword).toLowerCase())).length;
}

function average(values) {
  if (!values.length) return 0;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function main() {
  const args = process.argv.slice(2).filter((arg) => arg !== "--json");
  const targetPath = args[0] ?? "scripts/response_logs.json";
  const entries = JSON.parse(fs.readFileSync(targetPath, "utf8"));

  const uiCounter = {};
  const failureCounter = {};
  const failureDetails = {};
  const failureExamples = {};
  const personaScores = {};
  const groundingMissing = [];
  const grouped = new Map();
  const categoriesByKey = new Map();
  const perEntryFailures = [];

  for (const entry of entries) {
    const output = entry.output ?? "";
    const entryFailures = new Set();

    const uiFailures = uiBreaks(output);
    for (const failure of uiFailures) {
      uiCounter[failure] = (uiCounter[failure] ?? 0) + 1;
      const code = codeForFailure(failure);
      entryFailures.add(`${code}:${failure}`);
      addFailure(failureCounter, failureDetails, failureExamples, code, failure, entry);
    }

    for (const failure of entry.failures ?? []) {
      const code = codeForFailure(failure);
      entryFailures.add(`${code}:${failure}`);
      addFailure(failureCounter, failureDetails, failureExamples, code, failure, entry);
    }

    const mode = entry.mode ?? "unknown";
    personaScores[mode] ??= [];
    personaScores[mode].push(personaDensity(output));

    if ((entry.expectedGroundingKeywords ?? []).length && groundingHits(entry) === 0) {
      groundingMissing.push(entry.id ?? entry.prompt);
      entryFailures.add("RAG:missing expected grounding keyword");
      addFailure(failureCounter, failureDetails, failureExamples, "RAG", "missing expected grounding keyword", entry);
    }

    perEntryFailures.push({
      id: entry.id,
      mode,
      category: entry.category,
      prompt: entry.prompt,
      failures: [...entryFailures],
    });

    const key = JSON.stringify([entry.pass, entry.prompt]);
    const modes = grouped.get(key) ?? {};
    modes[mode] = normalizeAnswer(finalAnswer(output));
    grouped.set(key, modes);
    categoriesByKey.set(key, entry.category ?? "uncategorized");
  }

  const mismatches = [];
  const byCategory = {};
  const categoryTotals = {};

  for (const [key, modes] of grouped.entries()) {
    const pure = modes.pure;
    if (!pure) continue;

    for (const [mode, answer] of Object.entries(modes)) {
      if (mode === "pure" || !answer) continue;
      const category = categoriesByKey.get(key) ?? "uncategorized";
      categoryTotals[category] = (categoryTotals[category] ?? 0) + 1;
      if (answer !== pure) {
        mismatches.push({ key: JSON.parse(key), mode, pure, other: answer, category });
        byCategory[category] = (byCategory[category] ?? 0) + 1;
        addFailure(
          failureCounter,
          failureDetails,
          failureExamples,
          "DISC",
          "Nemanja/Pure Logic final answer mismatch",
          { mode, category, prompt: JSON.parse(key)[1] },
        );
      }
    }
  }

  const mismatchDenominator = Object.values(categoryTotals).reduce((sum, value) => sum + value, 0) || 1;
  const personaSummary = {};
  for (const [mode, scores] of Object.entries(personaScores)) {
    personaSummary[mode] = {
      count: scores.length,
      average: Number(average(scores).toFixed(3)),
      min: scores.length ? Math.min(...scores) : 0,
      max: scores.length ? Math.max(...scores) : 0,
    };
  }

  const failuresByCode = {};
  for (const code of Object.keys(FAILURE_TAXONOMY)) {
    if (!failureCounter[code]) continue;
    failuresByCode[code] = {
      label: FAILURE_TAXONOMY[code],
      count: failureCounter[code],
      details: failureDetails[code] ?? {},
      examples: failureExamples[code] ?? [],
    };
  }

  const requiresTargetedExpansion = Object.entries(categoryTotals)
    .filter(([category, total]) => total && ((byCategory[category] ?? 0) / total) > 0.03)
    .map(([category]) => category);

  let currentCleanStreak = 0;
  for (let index = perEntryFailures.length - 1; index >= 0; index--) {
    if (perEntryFailures[index].failures.length) break;
    currentCleanStreak++;
  }

  const saturationWindowSize = Math.min(250, perEntryFailures.length);
  const lastWindowFailures = perEntryFailures.slice(-saturationWindowSize).flatMap((entry) => entry.failures);
  const uniqueFailurePatternsLastWindow = [...new Set(lastWindowFailures)].sort();

  const summary = {
    entries: entries.length,
    ui_breaks: uiCounter,
    failure_taxonomy: FAILURE_TAXONOMY,
    failures_by_code: failuresByCode,
    mismatch_count: mismatches.length,
    mismatch_rate_percent: Number(((mismatches.length / mismatchDenominator) * 100).toFixed(3)),
    mismatch_by_category: byCategory,
    persona_density: personaSummary,
    grounding_missing_count: groundingMissing.length,
    grounding_missing_examples: groundingMissing.slice(0, 20),
    current_clean_streak_entries: currentCleanStreak,
    saturation_window_size: saturationWindowSize,
    unique_failure_patterns_last_window: uniqueFailurePatternsLastWindow,
    requires_targeted_expansion: requiresTargetedExpansion,
  };

  console.log(JSON.stringify(summary, null, 2));
  if (Object.keys(failuresByCode).length) {
    console.log(
      `\nAudit failures detected: ${Object.entries(failuresByCode)
        .map(([code, info]) => `${code}=${info.count}`)
        .join(", ")}`,
    );
    process.exitCode = 1;
  }
  if (requiresTargetedExpansion.length) {
    console.log(`\nTargeted expansion needed for: ${requiresTargetedExpansion.join(", ")}`);
    process.exitCode = 1;
  }
}

main();
